File: android/app/src/main/java/com/x1f4r/legioncontrol/agent/RemoteShell.kt
package com.x1f4r.legioncontrol.agent

/**
 * Which shell the far side hands an ssh command to, and how to write an argument for it.
 *
 * ssh does not take an argv. It takes one string and gives it to the account's login shell, so every
 * argument this app sends has to survive that shell's own reading of it. Which shell that is belongs
 * to the machine and cannot be guessed from the platform: OpenSSH on Windows runs cmd.exe unless the
 * DefaultShell registry value says otherwise, and plenty of installs set that to PowerShell.
 *
 * [AUTO] is what a configuration that has not said gets, and it is deliberately not a guess. While
 * every argument is one that means the same thing in all three shells, the command is written exactly
 * as it always was and nothing changes. The moment one of them would need quoting, the answer depends
 * on a shell nobody has named, and that is reported rather than resolved by picking one.
 */
enum class RemoteShell(val wire: String) {
    /** sh, bash, zsh, and anything else that reads a single-quoted string literally. */
    POSIX("posix"),

    /** cmd.exe, which is what OpenSSH on Windows starts with. */
    CMD("cmd"),

    /** PowerShell or pwsh, set as the account's DefaultShell. */
    POWERSHELL("powershell"),

    /** The configuration did not say. Safe only while nothing needs quoting. */
    AUTO("auto"),
    ;

    companion object {
        fun fromWire(value: String?): RemoteShell =
            entries.firstOrNull { it.wire.equals(value?.trim(), ignoreCase = true) } ?: AUTO
    }
}

/** An argument that cannot be written for the shell the far side is using, and why. */
class UnquotableArgument(val argument: String, val reason: String) :
    IllegalArgumentException(reason)

/**
 * The characters the contract's argv grammar allows: no spaces, no quotes, no dollar, no semicolon.
 *
 * An argument made only of these means the same thing to sh, cmd.exe and PowerShell, which is to say
 * nothing at all, so it is written exactly as it stands. That is also byte for byte what the previous
 * release sent, so every configuration that worked before produces the same command line now.
 *
 * The character set only, without the contract's "starts with a letter or a digit" rule. That rule
 * is about values this app *sends*, and it is enforced separately by [Contract.requireToken]; the
 * arguments checked here include absolute paths out of the configuration, which begin with a slash,
 * and `~/.legion-control/agent/src/index.mjs`, which begins with a tilde.
 *
 * `~` is in the set on purpose. A tilde inside quotes is a directory literally called `~` rather
 * than the home directory, so quoting it would break every install that uses the short form.
 */
private val BARE_SAFE = Regex("^[A-Za-z0-9._:@/\\\\~=+-]+$")

/**
 * cmd.exe reads these after CommandLineToArgvW has already finished, so no amount of double quoting
 * hides them: `%` still expands a variable and `!` still expands one under delayed expansion. They
 * can be escaped with a caret outside quotes, but not inside them, and every argument this app sends
 * that is long enough to need quoting is a path. Refusing is honest; a caret in the wrong place
 * silently changes the path.
 */
private val CMD_UNSAFE = charArrayOf('%', '!', '^', '&', '|', '<', '>', '\n', '\r')

/**
 * One command line for [shell], out of an argv that was never one.
 *
 * @throws UnquotableArgument when an argument cannot be written for that shell, which is a fault in
 *   the configuration rather than a fault at the far side, and is reported as such.
 */
fun buildRemoteCommand(shell: RemoteShell, arguments: List<String>): String {
    require(arguments.isNotEmpty()) { "a remote command needs at least one argument" }
    for (argument in arguments) {
        if (argument.isEmpty() && shell == RemoteShell.AUTO) {
            throw UnquotableArgument(
                argument,
                "An empty argument has to be quoted, and this system's configuration does not say " +
                    "which shell it logs in to. Add \"shell\": \"posix\", \"cmd\" or \"powershell\" " +
                    "to it.",
            )
        }
        if (argument.contains('\u0000')) {
            throw UnquotableArgument(argument, "An argument contains a null byte.")
        }
    }

    // Nothing needs quoting, so nothing is quoted. This is the case every existing configuration is
    // already in, and it produces byte for byte what this app has always sent.
    if (arguments.all { it.isNotEmpty() && BARE_SAFE.matches(it) }) {
        return arguments.joinToString(" ")
    }

    return when (shell) {
        RemoteShell.POSIX -> arguments.joinToString(" ") { quotePosix(it) }
        RemoteShell.CMD -> quoteCmdLine(arguments)
        RemoteShell.POWERSHELL -> quotePowerShellLine(arguments)
        RemoteShell.AUTO -> {
            val offender = arguments.first { !BARE_SAFE.matches(it) }
            throw UnquotableArgument(
                offender,
                "\"$offender\" has to be quoted, and this system's configuration does not say which " +
                    "shell it logs in to. Add \"shell\": \"posix\", \"cmd\" or \"powershell\" to the " +
                    "system so the argument can be written correctly.",
            )
        }
    }
}

/**
 * Single quotes, with the one escape sh has: close the quote, write an escaped quote, open it again.
 * Everything else inside single quotes is literal, including backslashes, dollars and newlines, so
 * this is total rather than a list of characters somebody has to keep up to date.
 */
internal fun quotePosix(argument: String): String {
    if (argument.isNotEmpty() && BARE_SAFE.matches(argument)) return argument
    return "'" + argument.replace("'", "'\\''") + "'"
}

/**
 * The Windows argv encoding, which is the reverse of what CommandLineToArgvW does.
 *
 * Backslashes are only special in front of a quote: a run of them before one is doubled and the
 * quote is escaped, a run of them at the end of a quoted argument is doubled so the closing quote is
 * not eaten, and everywhere else they stand for themselves. Getting this wrong is how
 * `C:\Program Files\` turns into a quote that never closes.
 */
internal fun quoteWindowsArgv(argument: String): String {
    if (argument.isNotEmpty() && BARE_SAFE.matches(argument)) return argument
    val out = StringBuilder("\"")
    var backslashes = 0
    for (character in argument) {
        when (character) {
            '\\' -> backslashes += 1
            '"' -> {
                out.append("\\".repeat(backslashes * 2 + 1))
                out.append('"')
                backslashes = 0
            }

            else -> {
                out.append("\\".repeat(backslashes))
                out.append(character)
                backslashes = 0
            }
        }
    }
    out.append("\\".repeat(backslashes * 2))
    out.append('"')
    return out.toString()
}

private fun quoteCmdLine(arguments: List<String>): String = arguments.joinToString(" ") { argument ->
    val offender = argument.firstOrNull { it in CMD_UNSAFE }
    if (offender != null) {
        throw UnquotableArgument(
            argument,
            "\"$argument\" contains '$offender', which cmd.exe expands even inside quotes. Move the " +
                "agent somewhere without it, or set \"shell\": \"powershell\" on this system if that " +
                "is what it actually logs in to.",
        )
    }
    quoteWindowsArgv(argument)
}

/**
 * PowerShell, where a single-quoted string is literal and an embedded quote is written twice.
 *
 * The call operator is what makes this a command rather than an expression. A line that begins with
 * a quoted string is a string literal in PowerShell and it prints the path instead of running it,
 * which is the one difference from cmd.exe that actually matters here.
 */
private fun quotePowerShellLine(arguments: List<String>): String {
    val quoted = arguments.map { "'" + it.replace("'", "''") + "'" }
    return "& " + quoted.joinToString(" ")
}

File: desktop/LegionControl.Desktop/Transport/RemoteShell.cs
using System.Text;

namespace LegionControl.Desktop.Transport;

/// Which shell the far side hands an ssh command to.
///
/// ssh does not run an argv. It joins whatever it was given with single spaces and hands the result
/// to the login shell on the far side, which does its own word splitting. So a path with a space in
/// it, or an id with a metacharacter in it, is broken unless the command is serialised here, once,
/// for the shell that will actually parse it.
public enum RemoteShell
{
    /// sh, bash, zsh, fish and everything else on Linux and macOS.
    Posix,
    /// Windows OpenSSH with DefaultShell left alone, which is cmd.exe.
    Cmd,
    /// Windows OpenSSH with DefaultShell set to PowerShell.
    PowerShell,
}

public static class RemoteShells
{
    /// What a system gets when its configuration does not name one. Existing setups were written
    /// against a Windows OpenSSH with the stock DefaultShell, so windows means cmd unless it says
    /// otherwise: changing that silently would break every machine already configured.
    public static RemoteShell Default(Platform platform) => platform switch
    {
        Platform.Windows => RemoteShell.Cmd,
        _ => RemoteShell.Posix,
    };

    public static RemoteShell? Parse(string? value) => value?.Trim().ToLowerInvariant() switch
    {
        "posix" or "sh" or "bash" or "zsh" => RemoteShell.Posix,
        "cmd" => RemoteShell.Cmd,
        "powershell" or "pwsh" => RemoteShell.PowerShell,
        _ => null,
    };

    /// One argv, quoted so the far side splits it back into exactly the same words.
    public static string Serialize(this RemoteShell shell, IReadOnlyList<string> argv)
    {
        switch (shell)
        {
            case RemoteShell.Posix:
                return string.Join(" ", argv.Select(PosixQuoted));
            case RemoteShell.Cmd:
                return string.Join(" ", argv.Select(CmdQuoted));
            case RemoteShell.PowerShell:
                // The call operator is needed the moment the command itself is quoted: PowerShell
                // reads a bare quoted string as a string to print, not as a program to run.
                if (argv.Count == 0) return "";
                var parts = argv.Select(PowerShellQuoted).ToList();
                return string.Join(" ", new[] { "&", parts[0] }.Concat(parts.Skip(1)));
            default:
                return string.Join(" ", argv);
        }
    }

    /// Single quotes, with the one escape a POSIX shell has: end the quoting, emit a literal quote,
    /// start it again. Everything inside single quotes is already literal, including newlines.
    public static string PosixQuoted(string value)
    {
        if (value.Length == 0) return "''";
        // Nothing to do for the overwhelmingly common case, and leaving it alone keeps the command
        // readable in a log.
        if (value.All(IsPosixSafe)) return value;
        return "'" + value.Replace("'", "'\\''") + "'";
    }

    /// PowerShell single quotes: literal throughout, and a quote of its own is doubled.
    public static string PowerShellQuoted(string value) => "'" + value.Replace("'", "''") + "'";

    /// cmd.exe, which has two layers: the C runtime's argv parsing inside the double quotes, and
    /// cmd's own metacharacter handling outside them.
    public static string CmdQuoted(string value)
    {
        var quoted = new StringBuilder("\"");
        var backslashes = 0;
        foreach (var character in value)
        {
            switch (character)
            {
                case '\\':
                    backslashes += 1;
                    quoted.Append(character);
                    break;
                case '"':
                    // Every backslash run immediately before a quote is doubled, then the quote is
                    // escaped. This is the rule the Windows C runtime parses by.
                    quoted.Append('\\', backslashes + 1);
                    quoted.Append('"');
                    backslashes = 0;
                    break;
                default:
                    backslashes = 0;
                    quoted.Append(character);
                    break;
            }
        }
        quoted.Append('\\', backslashes);
        quoted.Append('"');

        // cmd looks at the whole line before the runtime does, and it does not care that a
        // metacharacter is inside quotes when the line reaches it through ssh. Escaping them is
        // harmless when they are literal and necessary when they are not.
        var escaped = new StringBuilder(quoted.Length + 8);
        foreach (var character in quoted.ToString())
        {
            if ("^&|<>()!%".Contains(character)) escaped.Append('^');
            escaped.Append(character);
        }
        return escaped.ToString();
    }

    private static bool IsPosixSafe(char character) =>
        (character >= 'a' && character <= 'z')
        || (character >= 'A' && character <= 'Z')
        || (character >= '0' && character <= '9')
        || "._-/:=@+,".Contains(character);
}

public enum Platform
{
    Linux,
    Windows,
    Mac,
}

public static class Platforms
{
    public static Platform? Parse(string? value) => value?.Trim().ToLowerInvariant() switch
    {
        "linux" => Platform.Linux,
        "windows" or "win" => Platform.Windows,
        "mac" or "macos" or "darwin" => Platform.Mac,
        _ => null,
    };

    public static string DefaultName(this Platform platform) => platform switch
    {
        Platform.Linux => "Linux",
        Platform.Windows => "Windows",
        Platform.Mac => "macOS",
        _ => "System",
    };

    public static string Wire(this Platform platform) => platform switch
    {
        Platform.Linux => "linux",
        Platform.Windows => "windows",
        Platform.Mac => "mac",
        _ => "linux",
    };
}
